panic: bitlists are different lengths

[2020-02-16 04:30:52]  INFO sync: Requesting parent block currentSlot=267714 parentRoot=2fadb24ae0ae
panic: bitlists are different lengths

goroutine 191 [running]:
github.com/prysmaticlabs/go-bitfield.Bitlist.Overlaps(0xc03a8afb00, 0x11, 0x11, 0xc03b8e40c0, 0x11, 0x11, 0x7e)
	external/com_github_prysmaticlabs_go_bitfield/bitlist.go:148 +0x1cd
github.com/prysmaticlabs/prysm/beacon-chain/core/helpers.AggregateAttestations(0xc022090c00, 0x70, 0x80, 0x0, 0x1, 0x1, 0x0, 0x0)
	beacon-chain/core/helpers/attestation.go:35 +0x106
github.com/prysmaticlabs/prysm/beacon-chain/operations/attestations.(*Service).aggregateAttestations(0xc0112157c0, 0x1bc3d40, 0xc0000ba018, 0xc010ee0000, 0x9fd, 0xa00, 0x0, 0x0)
	beacon-chain/operations/attestations/aggregate.go:62 +0x5db
github.com/prysmaticlabs/prysm/beacon-chain/operations/attestations.(*Service).aggregateRoutine(0xc0112157c0)
	beacon-chain/operations/attestations/aggregate.go:28 +0x244
created by github.com/prysmaticlabs/prysm/beacon-chain/operations/attestations.(*Service).Start
	beacon-chain/operations/attestations/service.go:49 +0x61


Running sha256:a26a1ac18ae72875ec978cb04167910546c3452451592a87877a49ed275ace33

--- a/beacon-chain/core/helpers/attestation.go
+++ b/beacon-chain/core/helpers/attestation.go
@@ -16,6 +16,10 @@ var (
 	// ErrAttestationAggregationBitsOverlap is returned when two attestations aggregation
 	// bits overlap with each other.
 	ErrAttestationAggregationBitsOverlap = errors.New("overlapping aggregation bits")
+
+	// ErrAttestationAggregationBitsDifferentLen is returned when two attestation aggregation bits
+	// have different lengths.
+	ErrAttestationAggregationBitsDifferentLen = errors.New("different bitlist lengths")
 )
 
 // AggregateAttestations such that the minimal number of attestations are returned.
@@ -32,7 +36,7 @@ func AggregateAttestations(atts []*ethpb.Attestation) ([]*ethpb.Attestation, err
 		}
 		for j := i + 1; j < len(atts); j++ {
 			b := atts[j]
-			if !a.AggregationBits.Overlaps(b.AggregationBits) {
+			if a.AggregationBits.Len() == b.AggregationBits.Len() && !a.AggregationBits.Overlaps(b.AggregationBits) {
 				var err error
 				a, err = AggregateAttestation(a, b)
 				if err != nil {
@@ -50,6 +54,11 @@ func AggregateAttestations(atts []*ethpb.Attestation) ([]*ethpb.Attestation, err
 	for i, a := range atts {
 		for j := i + 1; j < len(atts); j++ {
 			b := atts[j]
+
+			if a.AggregationBits.Len() != b.AggregationBits.Len() {
+				continue
+			}
+
 			if a.AggregationBits.Contains(b.AggregationBits) {
 				// If b is fully contained in a, then b can be removed.
 				atts = append(atts[:j], atts[j+1:]...)
@@ -74,6 +83,9 @@ var signatureFromBytes = bls.SignatureFromBytes
 
 // AggregateAttestation aggregates attestations a1 and a2 together.
 func AggregateAttestation(a1 *ethpb.Attestation, a2 *ethpb.Attestation) (*ethpb.Attestation, error) {
+	if a1.AggregationBits.Len() != a2.AggregationBits.Len() {
+		return nil, ErrAttestationAggregationBitsDifferentLen
+	}
 	if a1.AggregationBits.Overlaps(a2.AggregationBits) {
 		return nil, ErrAttestationAggregationBitsOverlap
 	}

--- a/beacon-chain/core/helpers/attestation_test.go
+++ b/beacon-chain/core/helpers/attestation_test.go
@@ -184,13 +184,15 @@ func TestAggregateAttestations(t *testing.T) {
 			},
 		},
 		{
-			name: "two attestations with different bitlist lengths",
+			name: "attestations with different bitlist lengths",
 			inputs: []bitfield.Bitlist{
 				{0b00000011, 0b10},
+				{0b00000111, 0b100},
 				{0b00000100, 0b1},
 			},
 			want: []bitfield.Bitlist{
 				{0b00000011, 0b10},
+				{0b00000111, 0b100},
 				{0b00000100, 0b1},
 			},
 		},

--- a/beacon-chain/core/helpers/attestation_test.go
+++ b/beacon-chain/core/helpers/attestation_test.go
@@ -62,6 +62,22 @@ func TestAggregateAttestation_OverlapFails(t *testing.T) {
 	}
 }
 
+func TestAggregateAttestation_DiffLengthFails(t *testing.T) {
+	tests := []struct {
+		a1 *ethpb.Attestation
+		a2 *ethpb.Attestation
+	}{
+		{a1: &ethpb.Attestation{AggregationBits: bitfield.Bitlist{0x0F}},
+			a2: &ethpb.Attestation{AggregationBits: bitfield.Bitlist{0x11}}},
+	}
+	for _, tt := range tests {
+		_, err := helpers.AggregateAttestation(tt.a1, tt.a2)
+		if err != helpers.ErrAttestationAggregationBitsDifferentLen {
+			t.Error("Did not receive wanted error")
+		}
+	}
+}
+
 func bitlistWithAllBitsSet(length uint64) bitfield.Bitlist {
 	b := bitfield.NewBitlist(length)
 	for i := uint64(0); i < length; i++ {
@@ -165,6 +181,17 @@ func TestAggregateAttestations(t *testing.T) {
 			},
 			want: []bitfield.Bitlist{
 				{0b00000011, 0b1},
+			},
+		},
+		{
+			name: "two attestations with different bitlist lengths",
+			inputs: []bitfield.Bitlist{
+				{0b00000011, 0b10},
+				{0b00000100, 0b1},
+			},
+			want: []bitfield.Bitlist{
+				{0b00000011, 0b10},
+				{0b00000100, 0b1},
 			},
 		},
 	}

